### ISSUE TITLE
ignore-missing-import option does not work on false positive "Library stubs not installed for X"

### ISSUE THREAD

**Issue by alexanderdamiani** [python/mypy#10632]
**Bug Report**

I was recommended to post a new issue that is very close to [this one](https://github.com/python/mypy/issues/10601). I am specifically getting this error concerning the [`pyyaml`](https://pypi.org/project/PyYAML/) package.

When running 

```python
mypy . --ignore-imports
```

I am getting an error saying
```
main.py:1: error: Library stubs not installed for "yaml" (or incompatible with Python 3.9)
main.py:1: note: Hint: "python3 -m pip install types-PyYAML"
main.py:1: note: (or run "mypy --install-types" to install all missing stub packages)
main.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

**To Reproduce**

I've created a [repo demonstrating the issue](https://github.com/alexanderdamiani/mypy_error_example). To reproduce the error:

1. Go to the `Actions` tab
2. Select the `Python Windows mypy` workflow
3. Click `Run Workflow` and run it on the `main` branch

There are only three files:
1. `requirements.txt`
```
pyyaml
```
2. `main.py`
```python
import yaml
```
3. `action.yml` (GitHub Action workflow file)
    * builds a Docker container with `windows-latest` as the base image and runs `mypy . --ignore-missing-imports` on the top-level directory

**Expected Behavior**

* `--ignore-missing-imports` should ignore the error

**Actual Behavior**

* error is shown and there is no alternative flag to use to ignore it

**Your Environment**

- Mypy version used: 0.902
- Mypy command-line flags: `--ignore-missing-imports`
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.9.5
- Operating system and version: Docker base image `windows-latest`

**Comment by anders-kiaer** [comment]
We faced the same issue in our CI. Ended up adding `types-pyyaml` to our test requirements after reading [this blog post](http://mypy-lang.blogspot.com/2021/06/mypy-0900-released.html), more specifically:
> Note: Even if you are using --ignore-missing-imports, mypy will ask to install stubs for packages that had bundled stubs in previous mypy releases. This helps you avoid losing type checking precision from missing stubs when upgrading to mypy 0.900.

**Comment by jeffmm** [comment]
I thought the entire reason for this update was for consistent treatment of stubs for third party packages. So why was it decided to introduce a breaking change that causes `--ignore-missing-imports` to error inconsistently for third party packages?

**Comment by epwalsh** [comment]
I have a similar issue, but I'm not sure if it's exactly the same. I'm using `mypy==0.902` and I have both `requests` and `types-requests` installed. But I still get

```
error: Cannot find implementation or library stub for module named "requests"
```

when `mypy` is ran both with and without the `--ignore-missing-imports` option.

**Comment by JukkaL** [comment]
This error is a bit unfortunate, but it's by design. We made this error a little tricky to ignore, since otherwise many mypy users would silently lose type checking precision when upgrading to mypy 0.9xx: mypy would silently stop using various third-party library stubs that were previously bundled with mypy, resulting in a loss in type checking precision. By making the error hard to ignore, users at least will make an explicit decision to not install the type stubs.

If you don't want to install the `types-*` package, for some reason, the error can be ignored by having a module-level `ignore_missing_imports` setting in `mypy.ini`:

```
[mypy-yaml.*]
ignore_missing_imports = True
```

Usually I'd recommend against doing this, since type checking results would then be worse than with older mypy versions.

There's an idea about how to make installing type stubs easier: https://github.com/python/mypy/issues/10600#issuecomment-859655627 Maybe it would help with some of this? By using `--install-types --non-interactive` all these hard-to-ignore errors would then go away.

**Comment by JukkaL** [comment]
@epwalsh Your issue seems different. If you have `types-requests` installed, you shouldn't get that error. Can you open a separate issue about this?

**Comment by alexanderdamiani** [comment]
@JukkaL I am trying to use the flag `--install-types` but I am getting an error saying `Error: no mypy cache directory (you must enable incremental mode)`. I have tried adding flags to both turn off incremental mode and disable the cache

```python
mypy . --install-types --ignore-missing-imports --disallow-untyped-defs --disallow-incomplete-defs --show-error-codes --raise-exceptions --no-incremental --cache-dir=nul
```

but I am still getting the same error.

**Comment by JelleZijlstra** [comment]
You have to first run mypy in incremental mode, then run `--install-types`. This probably can be documented better.

**Comment by alexanderdamiani** [comment]
@JelleZijlstra If I'm running a CI pipeline though with GitHub Actions, if I run `mypy` first in incremental mode it will fail saying it's missing type stubs. Could you share your approach on how you handle it in your CI pipeline?

**Comment by JelleZijlstra** [comment]
I'm not running mypy in any pipeline like that, but if I were you I would first run mypy locally to get the list of stub packages I need, then put those in my requirements list (requirements.txt, setup.py, or whatever else you're using) so that they're available in CI.

**Comment by alexanderdamiani** [comment]
That works fine if I'm running `mypy` directly in the workflow. However, if `mypy` is being run as part of GitHub Action, it will fail to find the stub packages installed.

**Comment by JukkaL** [comment]
#10669 will support installing types and type checking in one step, by using `mypy --install-types --non-interactive src/`.

**Comment by hauntsaninja** [comment]
If there are changes mypy should make, please open a new issue. Fwiw, I recommend adding stubs to your test requirements files.

**Comment by paulaceccon** [comment]
> I have a similar issue, but I'm not sure if it's exactly the same. I'm using `mypy==0.902` and I have both `requests` and `types-requests` installed. But I still get
> 
> ```
> error: Cannot find implementation or library stub for module named "requests"
> ```
> 
> when `mypy` is ran both with and without the `--ignore-missing-imports` option.

I have the same issue and can't find a workaround or understand why I'm getting it.

**Comment by hauntsaninja** [comment]
I'm not able to reproduce, could you try upgrading to the latest version of `types-requests` (or possibly upgrading to mypy 0.931)?

I agree that this whole situation is pretty unfortunate and that `--ignore-missing-imports` would ideally just make this go away.

**Comment by hauntsaninja** [comment]
Are you sure you're installing types-requests in the right environment? I think you need to specify something like: https://pre-commit.com/#config-additional_dependencies
(I personally don't really recommend using pre-commit for mypy, since there are a number of pitfalls, but I get the desire to)

**Comment by paulaceccon** [comment]
Perfect, my bad @hauntsaninja. This solved it! 

```
-   repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.931
    hooks:
    -   id: mypy
        verbose: true
        args: [--show-error-codes]
        additional_dependencies: ['types-requests']
```

**Comment by dingobar** [comment]
Still having this error. The solution using `additional_dependencies` by @paulaceccon changed nothing.

My config looks like this:
```
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.931
    hooks:
      - id: mypy
        args: [--ignore-missing-imports, --show-error-codes]
```

... yet, when running the mypy hook, the following output is seen:

````
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

app/log.py:5: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
app/log.py:5: note: Hint: "python3 -m pip install types-PyYAML"
app/log.py:5: note: (or run "mypy --install-types" to install all missing stub packages)
app/log.py:5: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 32 source files)
```

The package `types-PyYAML` is installed in the environment.

My theory is that there is a bug, because this is only the case for the import errors where the "note" lines are logged. If I remove `--ignore-missing-imports`, I get many more errors, but not the one with the suggestion to install a package.

Anyone else facing this and if so any idea how to work around it?

**Comment by hauntsaninja** [comment]
`additional_dependencies: ['types-PyYAML']` should work.

There's a really shitty situation where some stub packages are special cased to ignore `ignore-missing-imports` for various reasons. https://github.com/python/mypy/issues/10619

If you can't get `types-PyYAML` to be installed in the right environment, you can always try:
```
[mypy-yaml.*]
ignore_missing_imports = True
```

**Comment by hauntsaninja** [comment]
Also I see mention of `__pypackages__`. I'm not sure where that's coming from, but PEP 582 is not accepted and mypy doesn't have support for it.

**Comment by elsiehupp** [comment]
I found my way here because mypy couldn't find `types-requests` when being invoked as a pre-commit hook. (`types-requests` was installed in both the user Python environment and the Poetry virtual environment, so this shouldn't have been happening, right?)

Anyway, this is the section from my `.pre-commit-config.yaml` that ended up working:

```yaml
-   repo: https://github.com/pre-commit/mirrors-mypy
    rev: v0.942
    hooks:
    -   id: mypy
        args: ["--install-types", "--non-interactive", "--ignore-missing-imports"]
```

I hope this is helpful!

**Comment by Garrett-R** [comment]
> you can always try:
> 
>     [mypy-yaml.*]
>     ignore_missing_imports = True

I believe this is for using the `mypy.ini` file.  For those using pyproject.toml, you can do this:

```toml
[[tool.mypy.overrides]]
module = "yaml"
ignore_missing_imports = true
```

where `module` can also be a comma-separated list.

**Comment by Garrett-R** [comment]
I'm also suggesting [here](https://github.com/python/mypy/issues/10619#issuecomment-1174208395) that this special-casing (ignoring the `ignore_missing_imports` for certain packages) be removed.

**Comment by KrishnanS2006** [comment]
> If you don't want to install the types-* package, for some reason, the error can be ignored by having a module-level ignore_missing_imports setting in mypy.ini:

I tried this, but it didn't work? Is the `mypy.ini` file not recognized by a `pre-commit` hook?
